### ISSUE TITLE
Connecting the Sunrise to an esp32 board

### ISSUE THREAD

**Issue by hapklaar** [Niclas73/Sunrise-master#1]
Hi,

Thanks for your work on this! I'm playing a bit with this sensor and am trying to get your examples to work, unfortunately haven't been able to get any output by using the i2c bus. UART works fine though. I'm new to i2c, so still learning :)

I'm thinking it's probably not working due to me not having any external pullup resistors on SCL and SDA. Was hoping the internal pullup function of wire.h would enable me to work without them. How did you connect the sensor to your board? Did you also connect VDDIO per the instructions in TDE7318.pdf (page 6)?

PS Got the device to show up on i2c bus with the proposed 10k pullup resistors, but its not able to communicate it seems. Output of serial is: Waiting for sunrise response from address 0x68

**Comment by Niclas73** [comment]
Have you tried i2c-scanner to verify the connections and the address... try that first. I can help you more tomorrow if it doesn’t work out 👍

Search for the i2c scanner if you don’t have it....

/Niclas

**Comment by Niclas73** [comment]
It could be due to settings of the sensor... have you tried the readInfo sketch? Are you reading continuously or just single requests?

**Comment by hapklaar** [comment]
i2c scanner finds the sensor and shows its at 0x68. Was also able to query the sensor over i2c successfully with another sketch only using wire.h. (see found code below). Unfortunately got the same waiting response with the readInfo sketch


```
#include <Wire.h>
char FIND_ADDRESS = 'F';
char READ_CO2 = 'R';
byte address = 0x68;
//byte address = 104; //slave address

void setup() {
// put your setup code here, to run once:
Wire.begin();
Wire.setClock(10000);
Serial.begin(115200);
Serial.println("Type F to find all devices on I2C bus and type R to read CO2 from S-11");
}

void sendDataAndReadToAddress(byte reg, byte address, int bufferSize, int readCount) {
Wire.beginTransmission(address);
for (int k = 0; k < bufferSize; k++) {
Wire.write(reg);
}
Wire.endTransmission(false);
delay(24);
Wire.requestFrom(address, readCount);
delay(12);
int nBytes = Wire.available();
byte buff[nBytes];
int i = 0;
while (Wire.peek() != -1) {
buff[i] = Wire.read();
Serial.print(buff[i],HEX);
Serial.print(" ");
i++;
}
//Serial.write(buff, nBytes);
Serial.print('\n');
int co2 = (buff[0] << 8) | buff[1];
Serial.print("CO2 Concentration: ");
Serial.print(co2);
Serial.print("\n");
}

void getAllAddress() {
//Serial.println("Looking for addresses");
byte error, address;
for (address = 1; address < 128; address++) {
Wire.beginTransmission(address);
error = Wire.endTransmission();
delay(5);
if (error == 0) {
Serial.print(address, HEX);
Serial.print(" ");
}
}
}

void loop() {
// put your main code here, to run repeatedly:
while (Serial.available() == 0) { }
//Delay 10ms to ensure we get the whole message
delay(15);
int serialBufferSize = Serial.available();
byte buff[serialBufferSize];
int i = 0;
while (Serial.peek() != -1 && i < serialBufferSize) {
buff[i] = (byte)Serial.read();
i++;
}
char command = (char)buff[0];
if (command == FIND_ADDRESS) {
getAllAddress();
} else if (command == READ_CO2) {
//The LSB nibble of the second byte contains the number of data bytes being read
//Add 2 for the Command byte and Checksum
int bufferSize = 1;
int readCount = 2;
sendDataAndReadToAddress(0x06, address, bufferSize, readCount);
} else {
Serial.println("Invalid Command... Type F to find all devices on I2C bus and type R to read CO2 from S-11");
}
}
```

**Comment by Niclas73** [comment]
Try to remove the following lines:

while(sunrise.readErrorStatus() >> 7){
    Serial.println("Waiting for first measurement to arrive...");  
    delay(1000);

I added it to avoid initial false readings...

**Comment by hapklaar** [comment]
The message I'm getting is "Waiting for sunrise response from address 0x68"

Not sure what removing that section of code would do as it's not happening in that loop?

**Comment by Niclas73** [comment]
Sorry, I didn’t read your error message careful enough. You get stuck during init so the i2c bus is not up. What pins are you using for SCL and SDA?

**Comment by Niclas73** [comment]
GPIO 22 (SCL) and GPIO 21 (SDA)?

**Comment by Niclas73** [comment]
GPIO 22 (SCL) and GPIO 21 (SDA)?

**Comment by hapklaar** [comment]
Yes indeed. GPIO21 and 22, which works with the sketch I posted. Which makes me guess the i2c bus is connected correctly to the sensor.

**Comment by Niclas73** [comment]
Yes...hmmm, strange. I use it in all my CO2 sensors without issue... I’ll check tomorrow so the latest code is pushed to my repo. Can’t think of anything else.

**Comment by hapklaar** [comment]
Great, thanks for taking the time!

In the mean time I broke the USB connector of the pcb, so got some soldering to do...

**Comment by Niclas73** [comment]
No problem!

**Comment by hapklaar** [comment]
PS just to be sure, have you tied up the sensor according to this diagram as well?

![image](https://user-images.githubusercontent.com/7703320/91746766-62f82f80-ebbd-11ea-8d10-74215d6c2819.png)

Only thing I left out is nRDY

**Comment by Niclas73** [comment]
If you plan to use it in continuous mode then this wiring is correct!

**Comment by Niclas73** [comment]
Mine are wired so they can use both single and continuous measurements, but you don’t necessarily have to do that if you plan to only use continuous mode...

**Comment by Niclas73** [comment]
![E5166311-EFE1-4124-8267-D68BCFE39159](https://user-images.githubusercontent.com/56290078/91747745-0138c500-ebbf-11ea-99a4-226f8b01260d.jpeg)

**Comment by Niclas73** [comment]
The ones I used as prototypes are wired as above. You can see the traces...

**Comment by Niclas73** [comment]
![AB24988C-9EC0-4C53-9DF3-25ACAC558250](https://user-images.githubusercontent.com/56290078/91748003-82905780-ebbf-11ea-8f17-0564b1bd8841.jpeg)

**Comment by hapklaar** [comment]
That's so cool! I like your pcb design, easy for mounting the sensor in an enclosure and limiting the number of wires needed to connect it especially.

I'm currently trying to use it for continuous usage, but also very interested in on demand reading in a battery powered unit. This sensor is great in that it doesn't require warmup time, so could be very useful for that.

Do you also use 10k pullup resistors, or are you using a lower value?

**Comment by Niclas73** [comment]
It is not that sensitive. I have successfully used from 4-10k... Yes, I have tested a lot of CO2 sensors and this one is in a league of its own. I wired up five and placed next to each other and they only differ 5-10 ppm after a few weeks in operation.

**Comment by Niclas73** [comment]
I used the sketch "continous.ino" as is in the library and it worked just fine here..?
![image](https://user-images.githubusercontent.com/56290078/91838725-422bea80-ec4e-11ea-875e-5d2e96722f50.png)

**Comment by hapklaar** [comment]
Curious... I'm sure I've also tried the continuous sketch and got that error. I will try again once I have my ESP32 board back up with the usb port attached ;)

**Comment by hapklaar** [comment]
Ok was able to test again, and got 2 different results with continuous v1.0.0. When I power on the ESP32 by connecting the USB, I get this:
![image](https://user-images.githubusercontent.com/7703320/91882620-871e4400-ec83-11ea-861b-b846ddabd997.png)

After I press the reset button, I get this:
![image](https://user-images.githubusercontent.com/7703320/91882686-a1f0b880-ec83-11ea-9f9c-3ca598337422.png)

Strange right?

**Comment by Niclas73** [comment]
In both cases the bus is down but for some reason you got past the first bus check in initialization in your first example..? Did you get “i2c ready!” in the beginning?

**Comment by hapklaar** [comment]
I can't see as I need to restart serial monitor for it to capture anything in that situation. I'm too late to see if that is the case.

By the way, when I disconnect SDA and SCL I get a different error:

Waiting for first measurement to arrive...
Failed to write to sensor!

I'm going to test some different boards. They are 8266's, I guess that is no problem?

**Comment by Niclas73** [comment]
Probably not the board since your other sketch works. You probably need some minor tweaks to run on 8266, data pins for example...

**Comment by hapklaar** [comment]
The other sketch is now also giving strange values. Must be something wrong with the board, wiring/breadboard or sensor.

How does you library determine SDA and SCL? Is it hardcoded or does it use wire.h designations?

**Comment by hapklaar** [comment]
Looking at all the compile errors, it looks like it's not gonna be easy enough for me. ;)

![image](https://user-images.githubusercontent.com/7703320/91885415-06157b80-ec88-11ea-86ba-ee4e2e90bc08.png)

**Comment by Niclas73** [comment]
Do you have a spare module or only one? Do you have an oscilloscope to check signals?

**Comment by hapklaar** [comment]
I don't have a spare ESP32 unfortunately, got a some Wemos D1 mini's, but they are 8266.

I do have a scope, but not very proficient with it yet.

**Comment by Niclas73** [comment]
Just put one probe on scl and one on sda and the crocodiles on ground. Start the scope an push “auto”. Set trigger level to 1.5V and trigger on rising edge. You should see a 400kHz square wave on scl and you should see the data packages on sda. You don’t have to dig any deeper than that. If you see “nice” signals going from 0V to 3.3 then you are all good. You can then see if you can see both the request package and the response package. If there are no packages att all then esp is faulty. If you see the request package but no response then the CO2 module is faulty... I’m sure you can do it! What kind of scope is it? If the signals are disturbed there could be a wiring issue or pull-up/pull-down problem... give it a try😀

**Comment by Niclas73** [comment]
https://www.allaboutcircuits.com/technical-articles/reading-the-i2c-bus-on-an-oscilloscope/

**Comment by hapklaar** [comment]
That sounds interesting and doable even for me 😀

I have a DSO138 and also have an old Tektronix 2215. Will give it a go when I have some more time, I hope later this week!
